### PR TITLE
Schedule commands in current thread context

### LINKED ISSUE #17143: Schedule commands in the current thread context
Right now `ThreadPool`'s `schedule` and `scheduleWithFixedDelay` schedule a command to run without any thread context. As of https://github.com/elastic/elasticsearch/pull/17077 it will be possible to intentionally schedule a command in the current thread context. We should make that behavior the default behavior.

### PR DESCRIPTION
**Pull request by ywelsch**
Changes ThreadPool's `schedule` method to run the schedule task in the context of the thread that scheduled the task.

This is the more sensible default for this method, and eliminates a range of bugs where the current thread context is mistakenly dropped.

Closes #17143

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java [hunk 1]
```diff
@@ -339,6 +337,7 @@ public ExecutorService executor(String name) {
      */
     @Override
     public ScheduledCancellable schedule(Runnable command, TimeValue delay, String executor) {
+        command = preserveContext(command);
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java]: Should `preserveContext` be private now?

**Review by nik9000 — COMMENTED** [review]
+1

It has been a while since I looked at this! The reason I never made this change when I worked on this before was that I thought that some calls to `schedule` *intentionally* didn't keep the thread context. I wonder if it'd be best to explicitly throw away the context in the places where we don't want to preserve it. It might be as simple as adding a boolean to `schedule` or something like that. I hate to change the signature but that'd *force* you to think about whether or not you want to preserve the context every time you call `schedule`.

**Review by jimczi — APPROVED** [review]
LGTM, it's hard to say if there are usages today that rely on the current default. Although scratching the context can make more harm than preserving it so I think we're on the right side with this change.

**Comment by ywelsch** [comment]
> It might be as simple as adding a boolean to schedule or something like that. I hate to change the signature but that'd force you to think about whether or not you want to preserve the context every time you call schedule

I thought about this initially, and looked at a lot of callers, coming to the conclusion that I couldn't find any place where not preserving the calling context would make sense. Adding a boolean would i think add more confusion, and can still be introduced if we have an actual case where we need it.

File: server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java (hunk 1 above)
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java]: I've removed the method from the Scheduler interface in 34a704d

**Review by nik9000 — APPROVED** [review]
LGTM then!

**Pull request merged**